### ISSUE TITLE
JSON should parse keys automatically, and not require hardcoded entries

### ISSUE THREAD

**Issue by ryanblock** [architect/parser#22]
**Describe the issue**
As the arc format evolves and both blessed and community-developed sections and settings are created, the JSON parser needs to be able to accommodate any inputs, not just that which is hardcoded into the JSON intermediary parser.

For example: https://github.com/architect/parser/blob/master/src/_json.js#L28-L35
```js
    if (section === 'aws') {
      result.aws = [
        ['region', json[section].region],
        ['profile', json[section].profile],
        ['runtime', json[section].runtime],
        ['layers', json[section].layers]
      ]
    }
```

Doesn't take into account `bucket` and other new things that have been added.

The same is true of JSON not supporting `@macros` (yet). The list will continue to go on forever, so we should parse dynamically, not with hardcoded keys.

**Comment by ryanblock** [comment]
See also: https://runkit.com/ryanblock/arc-parser---arc-vs-arc-json

**Comment by mikemaccana** [comment]
For my own reference:

```
var parser = require('@architect/parser'),
  fs = require('fs'),
  log = console.log.bind(log)

var arcJSON = fs.readFileSync('arc.json', 'utf-8')

var intermediate = parser.json(arcJSON)
var dotArc = parser.stringify(intermediate)

log(dotArc)
```

**Comment by mikemaccana** [comment]
We could easily have `_json()` (the json to intermediary) do a mapping of sections ('pragmas') to data types, but I don't quite understand why so many arrays are used in place of strings: 

```
app: [ 'appname' ],
```

or objects... 

```
  http:
   [ [ 'get', '/' ],
     [ 'get', '/login' ],
     [ 'get', '/signup' ],
     [ 'get', '/forgot' ],
     [ 'get', '/forgot/sent' ],
     [ 'get', '/logout' ],
```

in the intermediary format

**Comment by brianleroux** [comment]
because it is clean/fast and not intended to be something a human read/write but rather something a machine rips thru. its generic. the pragmas are keys. their values are always an array of values that are either scalar or compound types.

**Comment by mikemaccana** [comment]
> not intended to be something a human read/write but rather something a machine rips thru. 

Is performance an issue? Honestly feel it's worth optimising for people otherwise.

**Comment by brianleroux** [comment]
we did! the .arc format is for people. no syntax. discourages nesting. supports all the major scalar and compound types.

**Comment by mikemaccana** [comment]
I mean in all code in arc. Making it easy to develop for (knowing an jey: value is an object) vs optimising for the CPU.

**Comment by mikemaccana** [comment]
OK thing I've solved it, there seemed to be a couple of sections doing the same thing with slightly different code. Tests all pass.

**Comment by brianleroux** [comment]
I am intending to write a formal grammar and AST which is what I think you are asking?

**Comment by mikemaccana** [comment]
No just - maybe we have different backgrounds here - just using predictable data types. Like I'd expect 'age' to be a Number, I'd expect 'name' to be a String and I'd expect key: value pairs to be an Object rather than an array of 2 item arrays. It's not the 'what' in parser that I find confusing, it's the 'why'.

**Comment by brianleroux** [comment]
ok, see the impedance mismatch. The parser format has no specialized named keys that equal some specific type. The following is a valid .arc file:

```
@house
music
1
omg neat
```

that turns into `{balls: ['deep', 1, ['omg', 'neat']]}` with real numbers and everything. 

its the same thing with json, xml or yaml. they don't have a special meaning for `http` you create that meaning however you want. 

that being said of course we could create a higher level typed interface. app is a string. etc. that is no different than making a dialect on top of json or yaml too.

**Comment by brianleroux** [comment]
as to 'why' the reason is because we needed a small and **generic** file format that is terse, has comments and discourages nesting while supporting basic scalar types and compound types.

**Comment by mikemaccana** [comment]
OK I think this is done. I've pushed a cleaned up version to `master`:

 - Handing for different data types is based on the section/pragma name, with no knowledge of specific keys inside that section/pragma. So yes, you'd have to add `ses` to the right group if `ses` was added, but once that's done JSON doesn't have to know about each individual key beneath `ses`.

 - Tests now deep inspect the output and look for / show any deviations from what's expected, rather than just checking the `parse.json()` finished.

**Comment by brianleroux** [comment]
dope! thank u 💯 
lets keep talking about improving it; again the json support was pretty PoC so … idk we could look at adding typedefs or arc.js thing per https://github.com/architect/architect/issues/223

**Comment by mikemaccana** [comment]
Could you bump the version and release with these fixes when you get a mo? Doing `npm i https://github.com/architect/parser#7b51a01a6651f24eec3a78b2881456ddb64fa843` ATM. Thx!

**Comment by ryanblock** [comment]
Hasn't shipped to npm, but will be in Architect 6.0.14

**Comment by ryanblock** [comment]
Changes now live in parser `1.2.1`!

**Comment by ryanblock** [comment]
And released properly in Architect `6.0.14`, now out.

Thanks again @mikemaccana!
